Jetpack Manage: Pricing page - creating plan and product cards

coder-karen opened the pull request:
Related to https://github.com/Automattic/jetpack-manage/issues/247

## Proposed Changes

* Currently this PR creates plan cards for the new `manage/pricing` page, as well as product cards, Woo cards and Backup add-on cards.
* These are not yet connected to a license / bundle selector, so a figure of 1 is hard-coded in while developing.
* The savings line in each card currently says 'per bundle per month' for bundles, instead of per license (as the design specs suggest). This can be changed but I've left it as is for now.
* The Get button should directly open the license lightbox on the issue licenses page for the relevant product / including the relevant bundle size. The functionality is only for logged in users at the moment, and not for bundles larger than 1. 
* See the 'for later PRs' below for functionality that doesn't yet work as expected.



### For later PRs:

* The Get button functionality for non-logged in users needs to be added: https://github.com/Automattic/jetpack-manage/issues/259
* The Get button needs to work for bundles larger than 1: https://github.com/Automattic/jetpack-manage/issues/264
* Get buttons for multi select options lead to an error on first load only: https://github.com/Automattic/jetpack-manage/issues/267
* The 'More about' link (lightbox component) is still to be added: https://github.com/Automattic/jetpack-manage/issues/258
* Console errors exist when changing the plan type using the multi-select options (the Security plan card, VaultPress Backup product card): https://github.com/Automattic/jetpack-manage/issues/265
* Styling tweaks are needed to match design specs: https://github.com/Automattic/jetpack-manage/issues/266

## Testing Instructions

* To test this out, apply this PR and visit `http://jetpack.cloud.localhost:3000/manage/pricing`. You should see a Security 10GB and Complete card
* Try toggling between the 10GB and 1TB option in the Security plans card (and VaultPres Backup products card) - the content should change accordingly.
* You can change the bundleSize in the `LicenseProductsList` component (in `jetpack-cloud/sections/manage/pricing/primary/index.tsx`) to 5, and the product name in the Security card should change, and the prices in both cards. These will all match what you see on the `partner-portal/issue-licenses` page.
  * You may need to modify the bundleSize in `most-popular-plans.tsx` and `all-license-items.tsx` if the bundleSize isn't being picked up initially.
  * To test the 'Get' button, hover over any of the Get buttons and as a logged in partner you should see a URL similar to `https://cloud.jetpack.com/partner-portal/issue-license?product_slug=jetpack-anti-spam&bundle_size=5&source=manage-pricing-page`. As a logged in non partner the same button should now direct you to `https://cloud.jetpack.com/manage/signup?issue-license=yes&product_slug=jetpack-anti-spam&bundle_size=5&source=manage-pricing-page`.
    * The Get button link should also directly open up the 'review license' lightbox when directly clicking on it.
    * Note that this only works when the quantity is 1 (not yet setup for larger quantities - see follow up PRs) - nor on first load for variants (eg backup and security 10GB and 1TB).

~Edit to add - to be able to see plans and products, you'll need to make sure you have a partner account and also that a key is selected - you may need to visit and revisit the partner portal dashboard to make sure a key is selected and the sites dashboard loads as usual - then the pricing page will work. This is only while using the auth endpoint.~ This isn't needed anymore as we're using the public endpoint now.

Screenshot showing all product and plan card component additions from this PR:

<img width="452" alt="Screenshot 2024-01-31 at 12 24 10" src="https://github.com/Automattic/wp-calypso/assets/16754605/d047d350-6f08-407a-a62b-9b2d2bab762f">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

coder-karen committed:
Adding plan cards with plan prices an details

coder-karen committed:
Fixing some unit test failures, and modifying plans heading - adding subheading

coder-karen committed:
Adding hero images to the plan cards

coder-karen committed:
Adding cards for all other product types

coder-karen committed:
Removing a stray comment

tbradsha commented:
Rebased in hopes this would fix things (I saw some dependency changes), but no dice.

tbradsha committed:
Update import path

tbradsha committed:
Update another path

tbradsha commented:
As I understand it, we should be using an endpoint that doesn't require authentication; currently products don't attempt to load because `useProductsQuery()` only runs if `isPartnerOAuthTokenLoaded` is `true`, which means there's a partner associated (basically the same misery I had to run through with #86333). As such, on my end the only way it was loading was with all this extra stuff (I'd've commit it, but ultimately we don't want auth, so...).

```diff
diff --git a/client/jetpack-cloud/sections/manage/pricing/license-products-list/index.tsx b/client/jetpack-cloud/sections/manage/pricing/license-products-list/index.tsx
index 2979bb0fe8..dc685d27b9 100644
--- a/client/jetpack-cloud/sections/manage/pricing/license-products-list/index.tsx
+++ b/client/jetpack-cloud/sections/manage/pricing/license-products-list/index.tsx
@@ -1,5 +1,12 @@
 import { useTranslate } from 'i18n-calypso';
+import QueryJetpackPartnerPortalPartner from 'calypso/components/data/query-jetpack-partner-portal-partner';
+import QueryJetpackPartnerKey from 'calypso/components/data/query-jetpack-partner-portal-partner-key';
 import useProductAndPlans from 'calypso/jetpack-cloud/sections/partner-portal/primary/issue-license/licenses-form/hooks/use-product-and-plans';
+import { useSelector } from 'calypso/state';
+import {
+       getCurrentPartner,
+       hasJetpackPartnerAccess as hasJetpackPartnerAccessSelector,
+} from 'calypso/state/partner-portal/partner/selectors';
 import { AllLicenseItems } from './all-license-items';
 import { MostPopularPlans } from './most-popular-plans';
 
@@ -20,8 +27,13 @@ export const LicenseProductsList = ( { bundleSize }: ProductsListProps ) => {
                productSearchQuery,
        } );
 
+       const hasJetpackPartnerAccess = useSelector( hasJetpackPartnerAccessSelector );
+       const partner = useSelector( getCurrentPartner );
+
        return (
                <div className="jetpack-product-store__products-list">
+                       { hasJetpackPartnerAccess && ! partner && <QueryJetpackPartnerPortalPartner /> }
+                       { hasJetpackPartnerAccess && <QueryJetpackPartnerKey /> }
                        <MostPopularPlans
                                heading={ translate( 'Plans' ) }
                                subHeading={ translate(
```

tbradsha commented on client/jetpack-cloud/sections/manage/pricing/license-products-list/featured-license-multi-item-card.tsx:
```diff
@@ -0,0 +1,114 @@
+
+	const [ item, setItem ] = useState( items[ 0 ] );
+
+	let title = item.name;
+	if ( title.startsWith( 'Jetpack Security' ) ) {
+		title.replace( 'Jetpack Security', 'Security' );
+	}
+	if ( title.startsWith( 'Jetpack VaultPress' ) ) {
+		title.replace( 'Jetpack VaultPress', 'VaultPress Backup' );
+	} else {
+		title = getProductShortTitle( item, true );
+	}
```
Unless I'm missing something, I think this whole section can be simplified:

```suggestion
	let title = getProductShortTitle( item, true );
```

coder-karen replied:
Looks like it, yes. Seemed to need it earlier but I'm wondering if when testing titles initially the cards weren't rendering properly as I started experiencing that later (I had to manually add a bundleSize of eg 5 to the most-popular-plans file and the all-license-items file as the result using the bundleSize in the main index.tsx file was a mesh between multi and single product cards 😕 ).I imagined that when linking with license numbers via the selector later on, that we'd be making sure the correct state is being passed down then, but  I'll see what I uncover in the meantime.
Fixed this in this commit: e9dd3afd5dbbda87b34d96ab04efea54e203f064

tbradsha commented on client/jetpack-cloud/sections/manage/pricing/license-products-list/simple-license-multi-item-card.tsx:
```diff
@@ -0,0 +1,113 @@
+
+	const [ item, setItem ] = useState( items[ 0 ] );
+
+	let title = item.name;
+	if ( title.startsWith( 'Jetpack Security' ) ) {
+		title.replace( 'Jetpack Security', 'Security' );
+	}
+	if ( title.startsWith( 'Jetpack VaultPress' ) ) {
+		title = 'VaultPress Backup';
+	} else {
+		title = getProductShortTitle( item, true );
+	}
```
As above, probably can do with this alone:

```suggestion
	let title = getProductShortTitle( item, true );
```

coder-karen replied:
Done in e9dd3afd5dbbda87b34d96ab04efea54e203f064.

tbradsha committed:
Type typo

tbradsha committed:
Oops, typed a Typescript typo due to previous Typescript typo typed not being a typo

tbradsha committed:
More compliant fix

tbradsha committed:
Fix for other component as well

tbradsha commented on client/jetpack-cloud/sections/manage/pricing/license-products-list/featured-license-multi-item-card.tsx:
```diff
@@ -0,0 +1,114 @@
+				} )
+			);
+		},
+		[ dispatch, item.slug, items ]
```
No need to include this as a dependency or use it in Tracks, as it's the same as `selectedProduct.slug` and `selectedProductSlug`, right?

coder-karen replied:
Good point! Fixed in e9dd3afd5dbbda87b34d96ab04efea54e203f064.

tbradsha reviewed (commented):
I didn't do a thorough review, as it's still a draft, but overall things make sense. I fixed a few tests and had a few other comments to consider:

* This PR uses `item` and `items` to refer to different things, which gets rather confusing, particularly when passing between components the data is different. I'd stick with "products" or "variants" or something a bit more descriptive and consistent; this would've helped me debug a test a bit more efficiently. That said, this is also somewhat the nature of the existing code, so not a blocker.
* The default/initial product variant isn't selected.
* When switching between product variants, the first click changes the price but throws a JS error and doesn't show the radio button as selected: `Uncaught TypeError: Cannot read properties of undefined (reading 'type')`
* The second click shows the radio button as selected, but going back and forth between them behaves the same way.

coder-karen committed:
Simplifying title generation, and product slug retrieval for tracking

coder-karen committed:
Changing items name to variants, in the multi item card context - better reflects usage

coder-karen committed:
Fixing variant option selection visibility for multi choice cards

coder-karen commented:
Thank you @tbradsha !

> As I understand it, we should be using an endpoint that doesn't require authentication

Yes that's correct. I edited the PR description after your comment to what should be a much easier way to test for now (for me anyway), I'm sorry you were in a position to have to create a snippet to get this work in the meantime though. 

> This PR uses item and items to refer to different things, 

I see what you mean.

Variants fits where I've used items (to refer to an Array of APIProductFamily items, which is an array of variants) and products is usd here to refer to a specific type of items. So that's in the MultiItem cards.

I notice we use products as a type of 'item' (there is product, plan, etc). So Im not sure about modifying items elsewhere to product, maybe there is a better word :think: That said, I've used product elsewhere where it doesn't necessarily fit, such as LicenseProductsList.

I need to think more on that one!

> When switching between product variants, the first click changes the price but throws a JS error and doesn't show the radio button as selected: Uncaught TypeError: Cannot read properties of undefined (reading 'type')

Yeah, I mentioned that in the PR description as I'd spotted it but wasn't initially sure of it. I've looked into it more now, and resolved the issue where you need to click the variant twice (the MultipleChoiceQuestion 'name' now differs between the featured license multi item card, and simple license multi item card). This also resolves not having a default variant selected. 
But still haven't come up with a solution for the console error (it happens when calling the dispatch function). I will come back to this one a little later...

coder-karen committed:
Adding product icons to simple cards, as well as style tweaks helping correct spacing between product cards

coder-karen committed:
Adding Get button links for logged in users

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into add/jetpack-manage-pricing-plan-product-components

coder-karen committed:
Using new public products API endpoint to gather product and price data

coder-karen committed:
Making sure Get button links not only redirect but opwn up the review license lightbox as well

coder-karen committed:
Modifying backup icons

coder-karen committed:
Modifying Get button functionality so it works for product variations

coder-karen committed:
Modifying type change back to what it was to prevent issues in othe files

cleacos commented:
It works as expected!

- You should see a Security 10GB and Complete card => YES
- 
<img width="615" alt="image" src="https://github.com/Automattic/jetpack-manage/assets/9832440/af78f325-645f-4aad-a450-9e9383ca6d04">

Is the size for 10GB and 1TB correct? I don't see it in the design, but it seems small and aligned to top.

For the VaultPress Backup product card looks better:
<img width="579" alt="image" src="https://github.com/Automattic/wp-calypso/assets/9832440/11be683a-9929-4351-98d9-52c349903e14">

Also, I tested it by clicking on the `Get` buttons I had this strange price for Security 10GB, $758. But... now I cannot replicate it again, I got this price:

<img width="1037" alt="image" src="https://github.com/Automattic/wp-calypso/assets/9832440/4b190cf2-44b5-439e-b6a0-0064a747ab17">

and this was the link `partner-portal/issue-license?product_slug=jetpack-security-t1&source=manage-pricing-page`

and this for the Security 1TB:

<img width="1034" alt="image" src="https://github.com/Automattic/wp-calypso/assets/9832440/4fe3388e-994b-4b39-ade7-e42a0f5ff7ed">

But now I cannot replicate it again...

cleacos commented:
In the tab with that issue with the price, I clicked, and I got a blank page, so I checked the console log and saw this error:

![image](https://github.com/Automattic/wp-calypso/assets/9832440/889b409d-5a3e-44ab-98e7-ba9f972a569d)

Dropping here in case someone else gets the same error. I cannot replicate it again.

cleacos commented:
We have an issue with the currency variable... I cleared the data (cookies) and reload it again, and I got the same console error.

coder-karen commented:
Thanks for testing @cleacos !

> Is the size for 10GB and 1TB correct?

Probably not. I'll add it to the design specs issue here - https://github.com/Automattic/jetpack-manage/issues/266 -  though I suspect there is a lot more design-wise. I've tried to focus on a few more obvious things (eg. icon size), but there will be more needed.


> Also, I tested it by clicking on the Get buttons I had this strange price for Security 10GB, $758. But... now I cannot replicate it again, I got this price:

That's curious. If you can replicate it, we should add it to a new issue or one of the existing ones, for sure.

> In the tab with that issue with the price, I clicked, and I got a blank page (in the Pricing page), so I checked the console log and saw this error:

Yes, I was getting that console error a lot. That was happening when using multi select cards and then attempting to get the correct selected product on the issue licenses page via the query params. The issue seemed related to the security and backup options not being the expected types within the APIProductFamilyProduct object, and so there have been various workarounds as a result. Though I wasn't getting it recently, I'll do more testing now as if it's a quick fix it should be included here (otherwise a follow-up).

coder-karen commented:
For now I've created a follow-on issue for the currency console error (and the issue using the Get button from the multi choice option cards): https://github.com/Automattic/jetpack-manage/issues/267

coder-karen committed:
Merge branch 'trunk' into add/jetpack-manage-pricing-plan-product-components

tbradsha commented on client/jetpack-cloud/sections/manage/pricing/utils/get-api-family-product-icon.tsx:
```diff
@@ -0,0 +1,93 @@
+	} ),
+};
+
+const getAPIFamilyProductIcon = ( { productSlug, light }: productIconProps ): string => {
```
This looks quite similar to `client/my-sites/plans/jetpack-plans/product-store/utils/get-product-icon.ts`...in an ideal world, we'd avoid repeating ourselves.

coder-karen replied:
Originally I attempted to change this file directly, but found I was virtually doubling everything up to create a new PRODUCT_ICON_MAP const for this use case, and change getAPIFamilyProductIcon to accept a string as well as productIconProps, then check for either.  Possibly marginally less duplication by doing it this way - certainly less duplication on the imports and less code generally having two files rather than one 🤔

tbradsha commented on client/my-sites/plans/jetpack-plans/product-store/hero-image/index.tsx:
```diff
@@ -47,17 +47,29 @@ const HERO_IMAGES: Record< string, { lg: string; md: string; sm: string } > = {
+const HERO_IMAGES_API_FAMILY: Record< string, { lg: string; md: string; sm: string } > = {
+	[ 'jetpack-security-t1' ]: {
+		lg: SecurityLg,
+		md: SecurityLg,
```
For future reference: this is not a typo; there is no medium image at the moment. Ideally we'd note this as a comment in code.

tbradsha reviewed (approved):
I put a few comments inline. While I think there are some optimisations in code we could do, I didn't see any blockers here. Visual polish can be done in a follow-up PR.

For your notes after merge:
* Should variant selection be above or below the price? For plans it currently is above, and for products it is below.
* For Security 1TB and Backup 1TB, clicking Get leads to the "More Info" lightbox instead of the "Review license" lightbox.

coder-karen commented:
Brilliant, thank you @tbradsha .

> Should variant selection be above or below the price? For plans it currently is above, and for products it is below.

Should be above as per design specs. Added to the design cleanup issue: https://github.com/Automattic/jetpack-manage/issues/266

> For Security 1TB and Backup 1TB, clicking Get leads to the "More Info" lightbox instead of the "Review license" lightbox.

Ah, thanks for spotting that. I've added it to this issue: https://github.com/Automattic/jetpack-manage/issues/267

cleacos commented:
> I'll add it to the design specs issue here - https://github.com/Automattic/jetpack-manage/issues/266

Thank you!

> That's curious. If you can replicate it, we should add it to a new issue or one of the existing ones, for sure.

I'll try again, but I don't remember the combination of actions I did for it.

> I'll do more testing now as if it's a quick fix it should be included here (otherwise a follow-up).

Yes, let's create another follow-up PR

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11131690).
